gdal/swig/python/scripts/gdal2tiles.py

```diff
@@ -768,234 +1451,120 @@ def open_input(self):
-                    if in_nodata == [] and self.out_ds.RasterCount in [1, 3]:
-                        tempfilename = self.gettempfilename('-gdal2tiles.vrt')
-                        self.out_ds.GetDriver().CreateCopy(tempfilename, self.out_ds)
-                        # open as a text file
-                        s = open(tempfilename).read()
-                        # Add the warping options
-                        s = s.replace(
-                            "<BlockXSize>",
-                            """
-  <VRTRasterBand dataType="Byte" band="%i" subClass="VRTWarpedRasterBand">
-    <ColorInterp>Alpha</ColorInterp>
-  </VRTRasterBand>
-  <BlockXSize>
-                            """ % (self.out_ds.RasterCount + 1))
-                        s = s.replace(
-                            "</GDALWarpOptions>",
-                            """
-    <DstAlphaBand>%i</DstAlphaBand>
-  </GDALWarpOptions>
-                            """ % (self.out_ds.RasterCount + 1))
-                        s = s.replace(
-                            "</WorkingDataType>",
-                            """
-    </WorkingDataType>
-    <Option name="INIT_DEST">0</Option>
-                            """)
-                        # save the corrected VRT
-                        open(tempfilename, "w").write(s)
-                        # open by GDAL as self.out_ds
-                        self.out_ds = gdal.Open(tempfilename)
-                        # delete the temporary file
-                        os.unlink(tempfilename)
-
-                        if self.options.verbose:
-                            print("Modified -dstalpha warping result saved into 'tiles1.vrt'")
-                            open("tiles1.vrt", "w").write(s)
-                    s = '''
-                    '''
-
-            else:
-                self.error("Input file has unknown SRS.",
-                           "Use --s_srs ESPG:xyz (or similar) to provide source reference system.")
+            if not in_srs:
+                exit_with_error(
+                    "Input file has unknown SRS.",
+                    "Use --s_srs ESPG:xyz (or similar) to provide source reference system.")
+
+            if not has_georeference(self.input_dataset):
+                exit_with_error(
+                    "There is no georeference - neither affine transformation (worldfile) "
+                    "nor GCPs. You can generate only 'raster' profile tiles.",
+                    "Either gdal2tiles with parameter -p 'raster' or use another GIS "
+                    "software for georeference e.g. gdal_transform -gcp / -a_ullr / -a_srs"
+                )
+
+            if ((in_srs.ExportToProj4() != self.out_srs.ExportToProj4()) or
+                    (self.input_dataset.GetGCPCount() != 0)):
+                self.warped_input_dataset = reproject_dataset(
+                    self.input_dataset, in_srs, self.out_srs)
+
+                if in_nodata:
```

Comment:
I'm wondering if you couldn't avoid all those direct editing of the XML of the warped VRT by using gdal.Warp('out.vrt', in_dataset, options = '-of VRT other_options_here'), the librarified version of gdalwarp.